src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,405 @@
+homomorphism sending `x` to `f x / g x`. -/
+@[to_additive "If `f` and `g` are additive Freiman homomorphisms to an additive commutative group,
+then `f - g` is the additive Freiman homomorphism sending `x` to `f x - g x`"]
+instance : has_div (A →*[n] G) :=
+⟨λ f g, { to_fun := λ x, f x / g x,
+  map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+    by rw [prod_map_div, prod_map_div, map_prod_eq_map_prod f hsA htA hs ht h,
+           map_prod_eq_map_prod g hsA htA hs ht h] }⟩
+
+@[simp, to_additive] lemma div_apply (f g : A →*[n] G) (x : α) : (f / g) x = f x / g x := rfl
+
+@[simp, to_additive] lemma div_comp (f₁ f₂ : B →*[n] G) (g : A →*[n] β) {hf hf₁ hf₂} :
+  (f₁ / f₂).comp g hf = f₁.comp g hf₁ / f₂.comp g hf₂ :=
+ext $ λ x, rfl
+
+/-! ### Instances -/
+
+/-- `A →*[n] β` is a `comm_monoid`. -/
+@[to_additive "`α →+[n] β` is an `add_comm_monoid`."]
+instance : comm_monoid (A →*[n] β) :=
+{ mul := (*),
+  mul_assoc := λ a b c, by { ext, apply mul_assoc },
+  one := 1,
+  one_mul := λ a, by { ext, apply one_mul },
+  mul_one := λ a, by { ext, apply mul_one },
+  mul_comm :=  λ a b, by { ext, apply mul_comm },
+  npow := λ m f,
+  { to_fun := λ x, f x ^ m,
+    map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+      by rw [prod_map_pow_right, prod_map_pow_right, map_prod_eq_map_prod f hsA htA hs ht h] },
+  npow_zero' := λ f, by { ext x, exact pow_zero _ },
+  npow_succ' := λ n f, by { ext x, exact pow_succ _ _ } }
+
+/-- If `β` is a commutative group, then `A →*[n] β` is a commutative group too. -/
+@[to_additive "If `β` is an additive commutative group, then `A →*[n] β` is an additive commutative
+group too."]
+instance {β} [comm_group β] : comm_group (A →*[n] β) :=
+{ inv := has_inv.inv,
+  div := has_div.div,
+  div_eq_mul_inv := by { intros, ext, apply div_eq_mul_inv },
+  mul_left_inv := by { intros, ext, apply mul_left_inv },
+  zpow := λ n f, { to_fun := λ x, (f x) ^ n,
+    map_prod_eq_map_prod' := λ s t hsA htA hs ht h,
+      by rw [prod_map_zpow_right, prod_map_zpow_right, map_prod_eq_map_prod f hsA htA hs ht h] },
+  zpow_zero' := λ f, by { ext x, exact zpow_zero _ },
+  zpow_succ' := λ n f, by { ext x, simp_rw [zpow_of_nat, pow_succ, mul_apply, coe_mk] },
+  zpow_neg'  := λ n f, by { ext x, simp_rw [zpow_neg_succ_of_nat, zpow_coe_nat], refl },
+  ..freiman_hom.comm_monoid }
+
+end freiman_hom
+
+/-! ### Hom hierarchy -/
+
+--TODO: merge this with `monoid_hom_class.fun_like` once #9888 is merged
+@[to_additive]
+instance monoid_hom.fun_like : fun_like (α →* β) α (λ _, β) :=
+{ coe := coe_fn,
+  coe_injective' := λ f g h, by { cases f, cases g, cases h, refl } }
+
+--TODO: change to `monoid_hom_class F A β → freiman_hom_class F A β n` once #9888 is merged
+@[to_additive]
+instance monoid_hom.freiman_hom_class : freiman_hom_class (α →* β) A β n :=
+{ map_prod_eq_map_prod' := λ f s t _ _ _ _ h, by rw [←f.map_multiset_prod, h, f.map_multiset_prod] }
```

Comment:
```suggestion
/-- A monoid homomorphism is naturally a `freiman_hom` on its entire domain.

We can't leave the domain `A : set α` of the `freiman_hom` a free variable, since it wouldn't be inferrable.
-/
@[to_additive]
instance monoid_hom.freiman_hom_class : freiman_hom_class (α →* β) set.univ β n :=
{ map_prod_eq_map_prod' := λ f s t _ _ _ _ h, by rw [←f.map_multiset_prod, h, f.map_multiset_prod] }
```